### DISCUSSION TITLE
Meaning of turn restriction 'only_straight_on'

### DISCUSSION THREAD

**Discussion by easbar** [valhalla/valhalla#4764] [General]
This 'only_straight_on' restriction relation: https://www.openstreetmap.org/relation/14932796 has from- and to- members and two via(-way) members. I'm wondering how these restrictions are meant to work. I could not find a proper definition in the [restriction wiki](https://wiki.openstreetmap.org/wiki/Relation:restriction). So far I thought this restriction means that coming from the from-member the only possible route should be following the via-members to the to-member. But in this case turning right from Bürgerweide onto Burgstraße is [clearly allowed](https://www.mapillary.com/app/?pKey=1141652099666480&focus=photo&lat=53.558976651488&lng=10.038521654463&z=17&x=0.49409771765423055&y=0.5040885436462355&zoom=0). Do you also consider this a mapping error or am I not understanding the meaning of these kind of restrictions, yet?

**Comment by nilsnolde** [comment]
Na you’re right, must be a mapping error. Your understanding is the same as ours. Is GH implementing via-way turn restrictions? :)

**Comment by easbar** [comment]
Ok, thanks. Just to be clear: In my understanding an 'only' turn restriction with ways `a-b-c-d` means that all routes starting with `a` must go to `b`, then `c`, then `d`. Something like `a-e`, `a-b-f` or `a-b-c-g` is not allowed. Something like `h-b-e`, `h-b-c-i`, `j-c-g` is possible, though (the restriction does not affect any routes that do not start with a).
In contrast, a via-way 'no' turn restriction is only about (conditionally) restricting the last turn: Here `a-b-c-d` means we can do any maneuver, except turning from `c` onto `d` **if** we started with `a-b-c`. So `a-b-c-e` would be illegal, but `f-b-c-d`, `f-g-c-d`, `a-b-h` and `a-i` are all allowed.

> Is GH implementing via-way turn restrictions? :)

GraphHopper supports via-way turn restrictions with a single via way for some time now. The first release including this feature was 7.0 (the current is 9.1). But it is still missing support for turn restrictions with multiple via-ways, but I'll fix this as soon as I'm sure what they are supposed to mean :)

Btw last time I checked there was an awful lot of broken turn restrictions, many of which are caused by edits in iD. Did you notice this? I posted a related issue some time ago here: https://github.com/openstreetmap/iD/issues/9346

**Reply by kevinkreiser** [reply]
i think this is the relevant part of the wiki: https://wiki.openstreetmap.org/wiki/Relation:restriction#Mandatory_restriction

i think i dont share the same interpretation as you. you seem to be saying that as soon as you touch the 'from' in the relation you must follow all the 'via's and end at the 'to', for the 'only' type of restriction. my understanding/assumption of an 'only' restriction/relation would be that if you travel on the path starting at the from and going through all the vias then you must also traverse the 'to'. you could have broken off from the relation at any point before the last 'via' though and it would have been allowed. specifically in your example above where you say all of these would be disallowed:

`a-e` `a-b-f` `a-b-c-g`

i think it would be reasonable to interpret it as only the last one would actually be disallowed. the reason i think this iirc this is how we implemented it in valhalla (see "complex restrictions"). really though @gknisely implemented it so he knows the details for sure

**Reply by easbar** [reply]
> i think this is the relevant part of the wiki: https://wiki.openstreetmap.org/wiki/Relation:restriction#Mandatory_restriction

Ok, but unfortunately it does not mention 'only'-turn restriction with one or more via-way(s), does it?

> you seem to be saying that as soon as you touch the 'from' in the relation you must follow all the 'via's and end at the 'to', for the 'only' type of restriction. 

Yes, that is what I was trying to say. (let's call this definition A)

>  my understanding/assumption of an 'only' restriction/relation would be that if you travel on the path starting at the from and going through all the vias then you must also traverse the 'to' (I'll call this definition B here)

That's very interesting, because this is what I thought they mean at first, too. But then looking at different occurrences I started to believe they are rather used the way I described (definition A above). That's exactly what got me thinking and lead me to ask here :) I'll try to collect some examples from current OSM.

**Reply by easbar** [reply]
[This route](https://www.openstreetmap.org/directions?engine=fossgis_valhalla_car&route=37.94538%2C-122.05780%3B37.94530%2C-122.05781#map=19/37.94517/-122.05723) would do a simple (and likely illegal) left turn without turn restrictions, but Valhalla seems to prevent it, because of this [restriction](https://www.openstreetmap.org/relation/15397381). It seems like Valhalla uses what I called definition A above. I.e. it does not allow turning from the from way onto the first via way.

[Same here](https://www.openstreetmap.org/directions?engine=fossgis_valhalla_car&route=37.37369%2C-121.94349%3B37.37343%2C-121.94382#map=17/37.37010/-121.94261) where this [restriction](https://www.openstreetmap.org/relation/1967240) forbids turning right at the end of the 'from' way, but rather forces the route onto the 'via' way and onto the 'to' way from there.

**Reply by kevinkreiser** [reply]
we have some tests for this. perhaps the 'only' type works this way while the other non-only type works as i mentioned. i swore this is what @gknisely  and i discussed though it was ages ago haha

**Comment by easbar** [comment]
Yes, the 'no' type works like you said. You can (but do not have to) follow along the from- and via-ways and only finally turning onto the to-way is restricted (if you arrived there this way). That's what got me thinking the 'only' type works similarly at first as well. But now I am pretty confident the 'only' type means it isn't allowed to leave the chain of from-, via- and to-ways once we are on the from-way. Shall we add this to the Wiki (once we agree)?

**Comment by gknisely** [comment]
Only restrictions are implemented based on the wiki [page](https://wiki.openstreetmap.org/wiki/Relation:restriction).  `Only` means that you must traverse the elements as you described above.  

> An 'only' turn restriction with ways `a-b-c-d` means that all routes starting with `a` must go to `b`, then `c`, then `d`.

We apply/create restrictions in all other situations from the `from` It should be noted that the type really does not matter here (meaning `only_straight_on` vs `only_right_turn`)  What we look at is the fact that this is an `only` restriction and pay attention to the `from`, `vias` and `to` to build the restrictions in all other directions.

Some tests exist [here](https://github.com/valhalla/valhalla/blob/master/test/gurka/test_only_restrictions.cc)

**Comment by easbar** [comment]
Ok, thanks! Btw the link points to a private repository.

**Reply by gknisely** [reply]
> Ok, thanks! Btw the link points to a private repository.

Sorry.  Updated.  Let me know if you have more questions

**Reply by easbar** [reply]
Thanks, I think we fully agree on the definition now.
